pkg/nodeset/v1alpha1/types.go

```diff
@@ -0,0 +1,192 @@
+
+	// NodeSetController specifies a NodeSet controller to be in charge of this NodeSet
+	// If not specified, the default one that comes with the NodeController in NodeClass
+	// will be used.
+	// +optional
+	NodeSetController string `json:"nodeSetController,omityempty"`
+
+	// Replicas is the number of desired replicas.
+	Replicas int32
+
+	// The maximum number of Nodes that can be created/deleted at a time when
+	// reconciliation takes place. This can be an absolute number (ex. 3) or a percentage
+	// of total nodes at the time of reconciliation. The absolute number will be
+	// calculated from percentage by round up.
+	// By default, a value of 1 is used
+	// +optional
+	BatchSize intstr.IntOrString
+
+	// The maximum number of Nodes that can be created above the original number of
+	// Nodes during rolling-update. This can be an absolute number or a percentage of
+	// total nodes at the time of reconciliation.
+	// A default of 1 is used if not set.
+	// +optional
+	OverCapacity intstr.IntOrString
+
+	// Config contains free form structural data that overrides those in NodeClass.
+	// +optional
+	Config string `json:"config,omitempty"`
+}
+
+type NodeSetConditionType string
+
+const (
+	// This is added when the NodeSet controllers failed to create or delete a node during
+	// reconciliation.
+	NodeSetReplicaFailure NodeSetConditionType = "ReplicaFailure"
+)
+
+// NodeSetCondition describes the state of a NodeSet at a certain point.
+type NodeSetCondition struct {
+	// Type of NodeSet condition.
+	Type NodeSetConditionType
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus
+	// The last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time
+	// The reason for the condition's last transition.
+	// +optional
+	Reason string
+	// A human readable message indicating details about the transition.
+	// +optional
+	Message string
+}
+
+type NodeSetStatus struct {
+	// Replicas is the number of actual replicas.
+	Replicas int32
+
+	// The number of replicas with running status for this NodeSet.
+	// +optional
+	RunningReplicas int32
+
+	// ObservedGeneration is the most recent generation observed by the controller.
+	// +optional
+	ObservedGeneration int64
+
+	// Represents the latest available observations of the NodeSet's current state.
+	// +optional
+	Conditions []NodeSetCondition
+}
+
+type NodeSet struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object's metadata.
+	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of desired behaviour of the NodeSet.
+	// +optional
+	Spec NodeSetSpec `json:"spec,omitempty"`
+
+	// Most recently observed status of the NodeSet.
+	// This data may not be up to date.
+	// Populated by the system.
+	// Read-only.
+	// +optional
+	Status NodeSetStatus `json:"status,omitempty"`
```

Comment:
Other resources also have a non-pointer status. So it's fine. I was of the impression we had pointers elsewhere.